### ISSUE TITLE
Cull unnecessary tasks

### ISSUE THREAD

**Issue by mrocklin** [ContinuumIO/dask#2]
Consider the following dask

```Python
In [2]: def inc(x): return x + 1

In [3]: d = {'x': 1, 'y': (inc, 'x'), 'z': (inc, 'x'), 'out': (add, 'y', 10)}
```

![mydask](https://cloud.githubusercontent.com/assets/306380/5683162/703f7488-97da-11e4-82b4-85191b6e00c8.png)

If we go ahead and get out the `out` key

```Python
>>> from dask.threaded import get
>>> get(d, 'out')
```

We end up doing excess work computing `'z'`.

It would be nice to have a function that takes a dask and a key or list of keys and removes all tasks not necessary to compute those keys.

```Python
>>> cull(d, ['out'])
{'x': 1, 'y': (inc, 'x'), 'out': (add, 'y', 10)}
```

![mydask](https://cloud.githubusercontent.com/assets/306380/5683204/cae193a8-97da-11e4-8fa2-0a8040687f71.png)

**Comment by mrocklin** [comment]
Where should functions like these live?  `core.py`?  Some sort of `analysis.py`?

**Comment by talumbau** [comment]
`analysis.py` seems good. The answer here would be:

```
{ ancestor(out) for out in outs }
```
with `ancestor(out)` in the DAG sense, right?

**Comment by mrocklin** [comment]
Hrm, what does ancestor return?

The result should be another dictionary, so one might write it in a similar form to what you've written above as the following:

    {k: v for k, v in dsk.items() if depends_on(outs, k)}

For some definition (and very efficient implementation) of `depends_on`.

**Comment by talumbau** [comment]
`ancestor` would essentially be the items for which `depends_on(outs, k)` returns `True`. I like yours better since I was implying a union of results which likely isn't necessary. Going through everything in the graph is probably fine because it's unlikely you would have a huge graph with only a small subset of the nodes surviving the cull.

**Comment by mrocklin** [comment]
Ah, I see, you're saying that we should start a traversal from just the `out` nodes.  That seems like a good idea.

I think that a common use case for this optimization would be relevant is to compute some giant blocked array expression `X` and then only ask for a small corner of that array (e.g. the first ten rows)  `X[:10, :10, :10]`.

**Comment by eriknw** [comment]
I think a more generic function like `subgraph` (or `subdask`, etc) that takes a dask, a node or list of nodes, and boolean keywords `successors` (or `descendants`, etc) and `predecessors` (or `ancestors`, etc) would be useful.

We can still add `cull` to the API, which would call the generic function.  If we wanted to get fancy (which, generally, we shouldn't get fancy unless there is a good reason too), we could have the node or list of nodes as optional input.  If not provided, the default nodes will be the nodes with no children.

**Comment by eriknw** [comment]
Closed via #10.
